Sources/NetworkProtection/Recovery/FailureRecoveryHandler.swift

```diff
@@ -0,0 +1,198 @@
+                    to: lastConnectedServer,
+                    includedRoutes: includedRoutes,
+                    excludedRoutes: excludedRoutes,
+                    isKillSwitchEnabled: isKillSwitchEnabled
+                )
+                try await updateConfig(result)
+                eventHandler(.completed(.unhealthy))
+                await reassertingControl?.stopReasserting()
+            } catch {
+                switch error {
+                case FailureRecoveryError.noRecoveryNecessary:
+                    await reassertingControl?.stopReasserting()
+                    eventHandler(.completed(.healthy))
+                default:
+                    eventHandler(.failed(error))
+                    throw error
+                }
+            }
+        }
+    }
+
+    func stop() {
+        task = nil
+    }
+
+    private func makeRecoveryAttempt(
+        to lastConnectedServer: NetworkProtectionServer,
+        includedRoutes: [IPAddressRange],
+        excludedRoutes: [IPAddressRange],
+        isKillSwitchEnabled: Bool
+    ) async throws -> NetworkProtectionDeviceManagement.GenerateTunnelConfigResult {
+        let serverSelectionMethod: NetworkProtectionServerSelectionMethod = .failureRecovery(serverName: lastConnectedServer.serverName)
+        let configurationResult: NetworkProtectionDeviceManagement.GenerateTunnelConfigResult
+
+        configurationResult = try await deviceManager.generateTunnelConfiguration(
+            selectionMethod: serverSelectionMethod,
+            includedRoutes: includedRoutes,
+            excludedRoutes: excludedRoutes,
+            isKillSwitchEnabled: isKillSwitchEnabled,
+            regenerateKey: false
+        )
+        os_log("🟢 Failure recovery fetched new config.", log: .networkProtectionServerFailureRecoveryLog, type: .info)
+
+        let newServer = configurationResult.server
+
+        os_log(
+            "🟢 Failure recovery - originalServerName: %{public}s, newServerName: %{public}s, originalAllowedIPs: %{public}s, newAllowedIPs: %{public}s",
+            log: .networkProtection,
+            type: .info,
+            lastConnectedServer.serverName,
+            newServer.serverName,
+            String(describing: lastConnectedServer.allowedIPs),
+            String(describing: newServer.allowedIPs)
+        )
+
+        guard lastConnectedServer.shouldReplace(with: newServer) else {
+            os_log("🟢 Server failure recovery not necessary.", log: .networkProtectionServerFailureRecoveryLog, type: .info)
+            throw FailureRecoveryError.noRecoveryNecessary
+        }
+
+        return configurationResult
+    }
+
+    private func incrementalPeriodicChecks(
+        _ config: RetryConfig,
+        action: @escaping () async throws -> Void
+    ) async {
+        let task = Task(priority: .background) {
+            var currentDelay = config.initialDelay
+            var count = 0
+            repeat {
+                do {
```

Comment:
Since task cancellation is cooperative, right before the `do` I'd suggest doing:

```swift
try Task.checkCancellation()
```